03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.Rmd

```diff
@@ -0,0 +1,319 @@
+It is good to keep in mind that various downstream analyses may benefit from varied strategies at this step. 
+Use `?mapIds` to see more options or strategies.
+
+In the next chunk, we will run the `mapIds()` function and supply the `multiVals` argument with the `"list"` option in order to get a large list with all the mapped values found for each gene identifier.
+
+```{r}
+# Map Ensembl IDs to their associated Entrez IDs
+mapped_list <- mapIds(
+  org.Dr.eg.db, # Replace with annotation package for the organism relevant to your data
+  keys = df$Gene,
+  column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
+  keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+  multiVals = "list"
+)
+```
+
+## Explore gene ID conversion
+
+Now, let's take a look at our mapped object to see how the mapping went.
+
+```{r}
+# Let's use the `head()` function to take a preview at our mapped list
+head(mapped_list)
+```
+
+It looks like we have Entrez IDs that were successfully mapped to the Ensembl IDs we provided.
+However, the data is now in a `list` object, making it a little more difficult to explore.
+We are going to turn our list object into a data frame object in the next chunk.
+
+```{r}
+# Let's make our object a bit more manageable for exploration by turning it into a data frame using the `reshape2::melt()` function
+mapped_df <- mapped_list %>%
+  reshape2::melt(value.name = "Entrez") %>%
+  dplyr::rename("Ensembl" = "L1") # Here we are renaming the column holding the Ensembl IDs as good practice
+```
+
+Now let's take a peek at our data frame.
+
+```{r}
+head(mapped_df)
+```
+
+We can see that our data frame has a new column `Entrez`.
+Let's get a summary of the values returned in the `Entrez` column of our mapped data frame.
+
+```{r}
+# We can use the `summary()` function to get a better idea of the distribution of values in the `Entrez` column
+summary(mapped_df$Entrez)
+```
+
+There are 9088 NA's in our data frame, which means that 9088 out of the 31882 Ensembl IDs did not map to Entrez IDs.
+This is not the best rate, in our opinion, which means that it may be wise to try a different gene identifier type that may have a better mapping rate.
+Regardless, it is always good to be aware of how many genes you are potentially "losing" if you rely on this new gene identifier you've mapped to for downstream analyses.
+
+Now let's check to see how many genes we have that were mapped to multiple IDs.
+
+```{r}
+multi_mapped_df <- mapped_df %>%
+  # Let's group by the Ensembl ID's in the `Ensembl` column
+  dplyr::group_by(Ensembl) %>%
+  # Create a new variable containing the number of Entrez IDs mapped to each ID
+  dplyr::mutate(entrez_id_count = dplyr::n()) %>%
+  # Arrange by the genes with the highest number of Entrez IDs mapped
+  dplyr::arrange(desc(entrez_id_count)) %>%
+  # Join the remaining data from `df` using the Ensembl IDs
+  dplyr::inner_join(df, by = c("Ensembl" = "Gene"))
```

Comment:
If we go with either of the options I mentioned, we could drop the inner_join here OR we could keep it so its abundantly clear that multi-mappings all end up having the same values.

Reply:
I agree with both points here, however, I chose to drop it in the last commit because I do not think it is completely necessary here as they are able to see that multi-mappings have the same values in the `all_mapped_df` object (if we choose to drop that object, we may want to revisit this comment/thought).